### DISCUSSION TITLE
Using WriteTo with Async sink

### DISCUSSION THREAD

**Discussion by joshevo** [AnderssonPeter/Hangfire.Console.Extensions#20] [General]
Recently we have run into problems with using WriteTo Hangfire with Serilog Async sink combination (on postgresql).
There was multiple NpgsqlOperationInProgressException exceptions rised.
It took a whole day to match, that Async was reusing a connection to postgres and run on it same time when DbLock was removing.
TLDR it would be a good idea to add a comment to NOT use Async sink when using WriteTo: Hangfire :)

**Comment by jmbryan4** [comment]


**Reply by AnderssonPeter** [reply]
Hi I'm not sure `Instead of writing to all Serilog sinks during job execution, we can defer writing to async sinks until the job has completed — while still showing real-time logs in the Hangfire Dashboard Console` would be a good default setting as it would make it harder to debug when the log entries are delayed, also the timestamp of log entries would be off? And it would cause a huge spike in log entries when a job is done?
So while I would accept a PR that adds a note about not combining it with Async Sinks, I'm not 100% sure I would accept a delay of log enties would be accepted..

**Reply by AnderssonPeter** [reply]
@jmbryan4 Please write your own messages instead of using an LLM to do it...

**Comment by AnderssonPeter** [comment]
> Recently we have run into problems with using WriteTo Hangfire with Serilog Async sink combination (on postgresql). There was multiple NpgsqlOperationInProgressException exceptions rised. It took a whole day to match, that Async was reusing a connection to postgres and run on it same time when DbLock was removing. TLDR it would be a good idea to add a comment to NOT use Async sink when using WriteTo: Hangfire :)

If you provide a PR that modifies the readme and adds a note about it, then I will Merge it :)

**Comment by jmbryan4** [comment]
@AnderssonPeter Is there anything that could be done outside of a warning in the README to prevent user of from running into this issue? 

I do like the idea of updating the README to include a warning on not using Serilogs.Sinks.Async with Hangfire sink.
Maybe this could also include updating the suggested json config and/or equivalent code config for using an async sink with Hangfire.Console.Extensions.Serilog?

I'm looking to understand the root problem and have a healthy discussion on this. I think more consumers that use Serilogs.Sinks.Async (https://github.com/serilog/serilog-sinks-async) or potentially other Serilog sinks that do some async batch logging could run into this.
Is the root issue that Hangfire.Console.Extensions.Serilog uses AsyncLocal<PerformingContext> to associate the current job with log events?

I understand that the user expects logs to appear immediately on the Hangfire Console, so the use of an async wrapper would go against the expected sync expectation. 

Again, thanks for your work on this project!

**Reply by jmbryan4** [reply]
It looks like the [Serilog.Sinks.Postgresql.Alternative](https://github.com/serilog-contrib/Serilog.Sinks.Postgresql.Alternative) sink already sends log events in async batches internally and may not benefit from the async wrapper?
[source](https://github.com/serilog-contrib/Serilog.Sinks.Postgresql.Alternative/blob/master/src/Serilog.Sinks.Postgresql.Alternative/Sinks/PostgreSQL/PostgreSQLSink.cs)

<img width="2147" alt="image" src="https://github.com/user-attachments/assets/73556460-6a19-4bdd-b5af-4b01a0ecc47f" />

**Reply by AnderssonPeter** [reply]
> It looks like the [Serilog.Sinks.Postgresql.Alternative](https://github.com/serilog-contrib/Serilog.Sinks.Postgresql.Alternative) sink already sends log events in async batches internally and may not benefit from the async wrapper?
> [source](https://github.com/serilog-contrib/Serilog.Sinks.Postgresql.Alternative/blob/master/src/Serilog.Sinks.Postgresql.Alternative/Sinks/PostgreSQL/PostgreSQLSink.cs)
> 
> <img width="2147" alt="image" src="https://github.com/user-attachments/assets/73556460-6a19-4bdd-b5af-4b01a0ecc47f" />

I'm not sure I understand the issue, can't you just not use async with this logger? While you keep it on other loggers?

**Reply by jmbryan4** [reply]
yes, it can not be used. I am good with my issue.

I was hoping there could be an update to the README on why the Hangfire sink does not work with Serilogs.Sinks.Async ([serilog/serilog-sinks-async](https://github.com/serilog/serilog-sinks-async)) in addition to the warning.
